### ISSUE TITLE
Using HASHSET populating cache with 1milion records and getting timeout error

### ISSUE THREAD

**Issue by BhavinPatelNW** [StackExchange/StackExchange.Redis#180]
Hi Friends,
I am populating StackExchange.Redis cache with 1milion records with 60 columns, but after populating 1,50,000 records I am getting below errors.
Timeout performing HMSET RecordDetails154858, inst: 1, mgr: ExecuteSelect, queue: 12, qu=0, qs=12, qc=0, wr=0/0, in=0/0

Can anybody help me in the resolution or any good approach suggestion in the same.
Thanks in advance.

I am waiting for quick positive response.

Thanks ...

**Comment by mgravell** [comment]
Do you have an example of what your code is doing here? i.e. how much it is
batching at a time, vs pipelining, etc?

I also wonder if this is due to persistence on the redis server: if you
have it configured to persist due to data changes, then yes: that could
interrupt things.

The queue of 12 is pretty short, so it doesn't look like a catastrophic
backlog. I would start by using "slowlog get" to investigate recent
long-running server-side commands, and looking at the redis server log (it
writes to the log when persisting)

On 2 April 2015 at 09:15, BhavinPatelNW <notifications@github.com> wrote:

> Hi Friends,
> I am populating StackExchange.Redis cache with 1milion records with 60
> columns, but after populating 1,50,000 records I am getting below errors.
> Timeout performing HMSET RecordDetails154858, inst: 1, mgr: ExecuteSelect,
> queue: 12, qu=0, qs=12, qc=0, wr=0/0, in=0/0
>
> Can anybody help me in the resolution or any good approach suggestion in
> the same.
> Thanks in advance.
>
> I am waiting for quick positive response.
>
> Thanks ...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StackExchange/StackExchange.Redis/issues/180>.
>



-- 
Regards,

Marc

**Comment by BhavinPatelNW** [comment]
Hi Mgravell,

Thanks for quick response and providing the details.

We are not using batching or pipelining for the same. If you can provide us the example for the same then its a good for us. Below are the code currently I am using.
1. Get data from database and set it to cache.
               var options = ConfigurationOptions.Parse("localhost");
                options.SyncTimeout = 10000;
                using (var redis = ConnectionMultiplexer.Connect(options))
                {
                    var db = redis.GetDatabase();
  string cnstring = ConfigurationManager.AppSettings["ConString"].ToString();
                        SqlConnection cn = new SqlConnection(cnstring);
                        SqlCommand cmd = new SqlCommand("StoredProcedureName", cn);
                        cmd.CommandType = CommandType.StoredProcedure;                                             
                        cn.Open();
                        rdr = cmd.ExecuteReader(CommandBehavior.CloseConnection);
                        while (rdr.Read())
                        {
                            CustomeDetail customDetail = new CustomeDetail();
                            customDetail.CustomeDetail1 = Convert.ToString(rdr[0]);
                            customDetail.CustomeDetail2 = Convert.ToString(rdr[1]);
                            customDetail.CustomeDetail3 = Convert.ToString(rdr[2]);
                            customDetail.CustomeDetail4 = Convert.ToString(rdr[3]);
                            customDetail.CustomeDetail5 = Convert.ToString(rdr[4]);
                            customDetail.CustomeDetail6 = Convert.ToString(rdr[5]);
                            customDetail.CustomeDetail7 = Convert.ToString(rdr[6]);
                            customDetail.CustomeDetail8 = Convert.ToString(rdr[7]);
                            customDetail.CustomeDetail9 = Convert.ToString(rdr[8]);
                            customDetail.CustomeDetail10 = Convert.ToString(rdr[9]);
                            customDetail.CustomeDetail11 = Convert.ToString(rdr[10]);
                            customDetail.CustomeDetail12 = Convert.ToString(rdr[11]);
                            customDetail.CustomeDetail13 = Convert.ToString(rdr[12]);
                            customDetail.CustomeDetail14 = Convert.ToString(rdr[13]);
                            customDetail.CustomeDetail15 = Convert.ToString(rdr[14]);
                            customDetail.CustomeDetail16 = Convert.ToString(rdr[15]);
                            customDetail.CustomeDetail17 = Convert.ToString(rdr[16]);
                            customDetail.CustomeDetail18 = Convert.ToString(rdr[17]);
                            customDetail.CustomeDetail19 = Convert.ToString(rdr[18]);
                            customDetail.CustomeDetail20 = Convert.ToString(rdr[19]);
                            customDetail.CustomeDetail21 = Convert.ToString(rdr[20]);
                            customDetail.CustomeDetail22 = Convert.ToString(rdr[21]);
                            customDetail.CustomeDetail23 = Convert.ToString(rdr[22]);
                            customDetail.CustomeDetail24 = Convert.ToString(rdr[23]);
                            customDetail.CustomeDetail25 = Convert.ToString(rdr[24]);
                            customDetail.CustomeDetail26 = Convert.ToString(rdr[25]);
                            customDetail.CustomeDetail27 = Convert.ToString(rdr[26]);
                            customDetail.CustomeDetail28 = Convert.ToString(rdr[27]);
                            customDetail.CustomeDetail29 = Convert.ToString(rdr[28]);
                            customDetail.CustomeDetail30 = Convert.ToString(rdr[29]);
                            customDetail.CustomeDetail31 = Convert.ToString(rdr[30]);
                            customDetail.CustomeDetail32 = Convert.ToString(rdr[31]);
                            customDetail.CustomeDetail33 = Convert.ToString(rdr[32]);
                            customDetail.CustomeDetail34 = Convert.ToString(rdr[33]);
                            customDetail.CustomeDetail35 = Convert.ToString(rdr[34]);
                            customDetail.CustomeDetail36 = Convert.ToString(rdr[35]);
                            customDetail.CustomeDetail37 = Convert.ToString(rdr[36]);
                            customDetail.CustomeDetail38 = Convert.ToString(rdr[37]);
                            customDetail.CustomeDetail39 = Convert.ToString(rdr[38]);
                            customDetail.CustomeDetail40 = Convert.ToString(rdr[39]);
                            customDetail.CustomeDetail41 = Convert.ToString(rdr[40]);
                            customDetail.CustomeDetail42 = Convert.ToString(rdr[41]);
                            customDetail.CustomeDetail43 = Convert.ToString(rdr[42]);
                            customDetail.CustomeDetail44 = Convert.ToString(rdr[43]);
                            customDetail.CustomeDetail45 = Convert.ToString(rdr[44]);
                            customDetail.CustomeDetail46 = Convert.ToString(rdr[45]);
                            customDetail.CustomeDetail47 = Convert.ToString(rdr[46]);
                            customDetail.CustomeDetail48 = Convert.ToString(rdr[47]);
                            customDetail.CustomeDetail49 = Convert.ToString(rdr[48]);
                            customDetail.CustomeDetail50 = Convert.ToString(rdr[49]);
                            customDetail.CustomeDetail51 = Convert.ToString(rdr[50]);
                            customDetail.CustomeDetail52 = Convert.ToString(rdr[51]);
                            customDetail.CustomeDetail53 = Convert.ToString(rdr[52]);
                            customDetail.CustomeDetail54 = Convert.ToString(rdr[53]);
                            customDetail.CustomeDetail55 = Convert.ToString(rdr[54]);
                            customDetail.CustomeDetail56 = Convert.ToString(rdr[55]);
                            customDetail.CustomeDetail57 = Convert.ToString(rdr[56]);
                            customDetail.CustomeDetail58 = Convert.ToString(rdr[57]);
                            customDetail.CustomeDetail59 = Convert.ToString(rdr[58]);
                            customDetail.CustomeDetail60 = Convert.ToString(rdr[59]);

                            db.HashSet("CustomDetail" + i, customDetail.ToHashEntries());
                            i++;
                        }
                        rdr.Close();                    
                    }

2. Below are the function we are utilizing to Get and Set HashEntry.
 public static HashEntry[] ToHashEntries(this object obj)
    {
        PropertyInfo[] properties = obj.GetType().GetProperties();
        return properties.Select(property => new HashEntry(property.Name, property.GetValue(obj).ToString())).ToArray();
    }

 public static T ConvertFromRedis<T>(this HashEntry[] hashEntries)
    {
        PropertyInfo[] properties = typeof(T).GetProperties();
        var obj = Activator.CreateInstance(typeof(T));
        foreach (var property in properties)
        {
            HashEntry entry = hashEntries.FirstOrDefault(g => g.Name.ToString().Equals(property.Name));
            if (entry.Equals(new HashEntry())) continue;
            property.SetValue(obj, Convert.ChangeType(entry.Value.ToString(), property.PropertyType));
        }

        return (T)obj;
    }
}

Also, i would like to share that, to get that much data, i am also getting similar kind of error.

We are looking for more details and help us on the same.

Thanks again.

**Comment by mgravell** [comment]
And did you have any information on the persistence configuration? A server
stall due to a data-related "save" is an immediate possibility.

**Comment by BhavinPatelNW** [comment]
Hi Mgravel,
Nop, no idea about it.
Can you please provide me the more inform on all points.

**Comment by mgravell** [comment]
ok; a redis-server can be configured with a wide range of persistence
options, or can be completely in-memory and never touch the disk. If this
is a *hosted* server (or a cloud server), it may be hard to find out. If
this  is *your* redis server, somebody who configured the server should
know the persistance configutaion.

The good news is; if you run redis-cli, you should be able to find the
configuration by issuing:

    info persistence

The most interesting values here are any "*last*time_sec" that aren't -1,
because that tells us it is saving. Another thing to do is to look at the
configuration itself; here's the example/documentation:
https://raw.githubusercontent.com/antirez/redis/2.8/redis.conf

Note that the configuration file contains a "logfile" entry; if this is
blank, the server writes to stdout; so either in the log-file or the server
console, we can see the server log. The key point here is that the log
includes persistence entries - for example here's the consequences of a
"bgsave":

[4740] 03 Apr 09:46:34.132 # fork operation complete
[4740] 03 Apr 09:46:34.464 * Background saving terminated with success

and a "save":

[4740] 03 Apr 09:46:39.672 * DB saved on disk

In all cases, what we're looking for is evidence of persistence at / around
the time of the blip.

Now; so far in the above I've focused on persistence because I think that
is the most likely culprit. However, other things could be in play; another
thing to do would be, in redis-cli, to run:

    slowlog get 50

the results of this can be hard to parse, but the second value in each
chunk is the unix timestamp, and the third is the number of microseconds
executing. If you don't want to have to parse it, the API is available in
SE.Redis via GetServer(...).SlowlogGet() - and does all the parsing etc for
you.

The thing we're looking for here is whether there were any long running
operations around the time of the blip that would have caused a blockage.


On 3 April 2015 at 08:16, BhavinPatelNW <notifications@github.com> wrote:

> Hi Mgravel,
> Nop, no idea about it.
> Can you please provide me the more inform on all points.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StackExchange/StackExchange.Redis/issues/180#issuecomment-89199412>
> .
>



-- 
Regards,

Marc

**Comment by mgravell** [comment]
For a concrete example of what the log looks like when there is a save, see
this example: https://groups.google.com/forum/#!topic/redis-db/96IDqxBEpsw

On 3 April 2015 at 09:52, Marc Gravell <marc.gravell@gmail.com> wrote:

> ok; a redis-server can be configured with a wide range of persistence
> options, or can be completely in-memory and never touch the disk. If this
> is a *hosted* server (or a cloud server), it may be hard to find out. If
> this  is *your* redis server, somebody who configured the server should
> know the persistance configutaion.
>
> The good news is; if you run redis-cli, you should be able to find the
> configuration by issuing:
>
>     info persistence
>
> The most interesting values here are any "*last*time_sec" that aren't -1,
> because that tells us it is saving. Another thing to do is to look at the
> configuration itself; here's the example/documentation:
> https://raw.githubusercontent.com/antirez/redis/2.8/redis.conf
>
> Note that the configuration file contains a "logfile" entry; if this is
> blank, the server writes to stdout; so either in the log-file or the server
> console, we can see the server log. The key point here is that the log
> includes persistence entries - for example here's the consequences of a
> "bgsave":
>
> [4740] 03 Apr 09:46:34.132 # fork operation complete
> [4740] 03 Apr 09:46:34.464 * Background saving terminated with success
>
> and a "save":
>
> [4740] 03 Apr 09:46:39.672 * DB saved on disk
>
> In all cases, what we're looking for is evidence of persistence at /
> around the time of the blip.
>
> Now; so far in the above I've focused on persistence because I think that
> is the most likely culprit. However, other things could be in play; another
> thing to do would be, in redis-cli, to run:
>
>     slowlog get 50
>
> the results of this can be hard to parse, but the second value in each
> chunk is the unix timestamp, and the third is the number of microseconds
> executing. If you don't want to have to parse it, the API is available in
> SE.Redis via GetServer(...).SlowlogGet() - and does all the parsing etc for
> you.
>
> The thing we're looking for here is whether there were any long running
> operations around the time of the blip that would have caused a blockage.
>
>
> On 3 April 2015 at 08:16, BhavinPatelNW <notifications@github.com> wrote:
>
>> Hi Mgravel,
>> Nop, no idea about it.
>> Can you please provide me the more inform on all points.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/StackExchange/StackExchange.Redis/issues/180#issuecomment-89199412>
>> .
>>
>
>
>
> --
> Regards,
>
> Marc
>



-- 
Regards,

Marc

**Comment by BhavinPatelNW** [comment]
Hi Mgravell,

Thanks for providing detail explanation.

Bit understand of this thing, but if you can provide the sample code in C# .net then its more understandable.

Also, I could not able to access below shared link. Shows connection fail error.
https://groups.google.com/forum/#!topic/redis-db/96IDqxBEpsw

Thanks again....

**Comment by BhavinPatelNW** [comment]
Hi Mgravell,

I am populating Redis Cache using stringset method through C# .net application. But I have observed that for some of the keys value not cached by redis. Can you please let us know how can we know which key value is not cached, so we can again cached it by writing logic.

Also, I have observed that using "Info" command shows "keyspace_hits" and "keyspace_misses" having values. How can we make it complete?

Please guide us on above points we are awaiting your valuable response.

Thanks in advance.

**Comment by mgravell** [comment]
That depends on a lot of things; is it possible, for example, that your
redis server is running out of memory and is evicting live keys based on
your configured key eviction policy? (at the redis server, not the client).
I'm not sure the  hit/miss numbers are relevant here.

**Comment by BhavinPatelNW** [comment]
Hi Mgravell,

Thanks for quick response.

Just want share that, Server have 4GB RAM. From that still 2GB RAM is available.Still such kind of issues I have observed. Also, to populate Redis cache I have used "Batch" approach.

**Comment by BhavinPatelNW** [comment]
Hi Mgravell,

I am providing the screenshot of server memory utilization and consumption by Redis. Also, to avoid some key are missing after cached it. Also, I am providing the Redis config file screenshot. Please let us know what is the issue behind it and how can we resolved it. We are awaiting your valuable response for the same. 

Also, please let us know if you require other information for the same.
![servermemoryinfo](https://cloud.githubusercontent.com/assets/11751873/7535119/49602808-f59f-11e4-8838-583a1b2f449e.png)
![rediscacheinfo](https://cloud.githubusercontent.com/assets/11751873/7535121/4962a34e-f59f-11e4-94cf-b31bff2a930c.png)
![redisconfigurationfile](https://cloud.githubusercontent.com/assets/11751873/7535122/49636086-f59f-11e4-8762-e43bcf0e68da.png)

We are using 1.0.414 version of StackExchange.Redis with targetframework="net45".

Again Thanks in advance and awaiting your valuable inputs on the same.

**Comment by abhit011** [comment]
I am also facing the similar kind of issue where I am writing around 50000 odd Keys and its respective values in a loop... And when i see the no of keys that were saved on Redis.Server was just not even 1000. I have disabled the entire persistence, by disabling the "save" and given maxmemory 268435456 . as 256 MB.. Then what wrong I am doing.. Please guide me ..

**Comment by abhit011** [comment]
When I exand the Task.WhenAll(results.ToArray()) , the keys from 0 till 580 shows status as "Ran To Completion" but rest 581 onwards till 50000 it shows "Waiting for activation" ?? Why  ?

**Comment by NickCraver** [comment]
We changed the behavior quite a bit here in 1.2.6. If anyone's still having the same issue with that release, please poke me here and I'll re-open. An update to latest should either: resolve your issue, or give a *much* more informative error message and guidance (it'd help to open an issue with the exact message).
